core/src/avm2/names.rs

```diff
@@ -0,0 +1,335 @@
+            AbcNamespace::Namespace(idx) => {
+                Self::Namespace(translation_unit.pool_string(idx.0, mc)?)
+            }
+            AbcNamespace::Package(idx) => Self::Package(translation_unit.pool_string(idx.0, mc)?),
+            AbcNamespace::PackageInternal(idx) => {
+                Self::PackageInternal(translation_unit.pool_string(idx.0, mc)?)
+            }
+            AbcNamespace::Protected(idx) => {
+                Self::Protected(translation_unit.pool_string(idx.0, mc)?)
+            }
+            AbcNamespace::Explicit(idx) => Self::Explicit(translation_unit.pool_string(idx.0, mc)?),
+            AbcNamespace::StaticProtected(idx) => {
+                Self::StaticProtected(translation_unit.pool_string(idx.0, mc)?)
+            }
+            AbcNamespace::Private(idx) => Self::Private(translation_unit.pool_string(idx.0, mc)?),
+        })
+    }
+
+    pub fn public_namespace() -> Self {
+        Namespace::Package("".into())
+    }
+
+    pub fn as3_namespace() -> Self {
+        Namespace::Namespace("http://adobe.com/AS3/2006/builtin".into())
+    }
+
+    pub fn package(package_name: impl Into<AvmString<'gc>>) -> Self {
+        Namespace::Package(package_name.into())
+    }
+
+    pub fn is_any(&self) -> bool {
+        match self {
+            Self::Any => true,
+            _ => false,
+        }
+    }
+
+    pub fn is_private(&self) -> bool {
+        match self {
+            Self::Private(_) => true,
+            _ => false,
+        }
+    }
+}
+
+/// A `QName`, likely "qualified name", consists of a namespace and name string.
+///
+/// This is technically interchangeable with `xml::XMLName`, as they both
+/// implement `QName`; however, AVM2 and XML have separate representations.
+///
+/// A property cannot be retrieved or set without first being resolved into a
+/// `QName`. All other forms of names and multinames are either versions of
+/// `QName` with unspecified parameters, or multiple names to be checked in
+/// order.
+#[derive(Clone, Collect, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[collect(no_drop)]
+pub struct QName<'gc> {
+    ns: Namespace<'gc>,
+    name: AvmString<'gc>,
+}
+
+impl<'gc> QName<'gc> {
+    pub fn new(ns: Namespace<'gc>, name: impl Into<AvmString<'gc>>) -> Self {
+        Self {
+            ns,
+            name: name.into(),
+        }
+    }
+
+    pub fn dynamic_name(local_part: impl Into<AvmString<'gc>>) -> Self {
+        Self {
+            ns: Namespace::public_namespace(),
+            name: local_part.into(),
+        }
+    }
+
+    /// Pull a `QName` from the multiname pool.
+    ///
+    /// This function returns an Err if the multiname does not exist or is not
+    /// a `QName`.
+    pub fn from_abc_multiname(
+        translation_unit: TranslationUnit<'gc>,
+        multiname_index: Index<AbcMultiname>,
+        mc: MutationContext<'gc, '_>,
+    ) -> Result<Self, Error> {
+        let actual_index = multiname_index.0 as usize - 1;
```

Comment:
This method is only called to resolve names during trait instantiation, so this should generate a runtime error rather than an underflow panic. I will change it as such.